// thirdweb/dist/cjs/react/web/ui/hooks/useWalletInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useWalletInfo = useWalletInfo;
exports.useWalletImage = useWalletImage;
const react_query_1 = require("@tanstack/react-query");
const getWalletInfo_js_1 = require("../../../../wallets/__generated__/getWalletInfo.js");
/**
 * Returns the wallet info for the provided wallet id.
 * @wallet
 */
function useWalletInfo(id) {
    return (0, react_query_1.useQuery)({
        queryKey: ["wallet-info", id],
        queryFn: () => {
            return (0, getWalletInfo_js_1.getWalletInfo)(id, false);
        },
        retry: false,
        refetchOnWindowFocus: false,
        refetchOnMount: false,
    });
}
/**
 * Returns the wallet image for the provided wallet id.
 * @wallet
 */
function useWalletImage(id) {
    return (0, react_query_1.useQuery)({
        queryKey: ["wallet-image", id],
        queryFn: () => {
            return (0, getWalletInfo_js_1.getWalletInfo)(id, true);
        },
        retry: false,
        refetchOnWindowFocus: false,
        refetchOnMount: false,
    });
}
